daedaland/AthanasiaOtc.sol

```diff
@@ -0,0 +1,811 @@
+}
+
+contract AthanasiaOtc is IAthanasiaOtc, Ownable, Pausable {
+    using SafeERC20 for IERC20;
+
+    uint256 constant PRICE_MULTIPLIER = 10**9;
+
+    IERC20 public immutable HEC;
+    IERC20 public immutable sHEC;
+    IStakingHelper public immutable stakingHelper;
+    IHectorTreasury public immutable hectorTreasury;
+
+    mapping(address => TokenInfo) collections;
+    mapping(address => mapping(address => uint256)) userPurchasedForCollection;
+
+    constructor(
+        address _HEC,
+        address _sHEC,
+        address _stakingHelper,
+        address _hectorTreasury
+    ) {
+        require(_HEC != address(0), 'OTC: Invalid HEC');
+        require(_sHEC != address(0), 'OTC: Invalid sHEC');
+        require(_stakingHelper != address(0), 'OTC: Invalid staking helper');
+        require(_hectorTreasury != address(0), 'OTC: Invalid hector treasury');
+
+        HEC = IERC20(_HEC);
+        sHEC = IERC20(_sHEC);
+        stakingHelper = IStakingHelper(_stakingHelper);
+        hectorTreasury = IHectorTreasury(_hectorTreasury);
+    }
+
+    modifier onlyExistingCollection(address collection) {
+        require(
+            collections[collection].totalAmount > 0,
+            'OTC: Unregistered collection'
+        );
+        _;
+    }
+
+    function registerCollection(
+        address collection,
+        address otcToken,
+        uint256 otcPrice,
+        uint256 totalAmount
+    ) external override onlyOwner {
+        require(collection != address(0), 'OTC: Invalid collection');
+        require(otcPrice > 0, 'OTC: Invalid otc price');
+        require(totalAmount > 0, 'OTC: Invalid total amount');
+
+        TokenInfo storage info = collections[collection];
+        info.otcToken = otcToken;
+        info.otcPrice = otcPrice;
+        info.totalAmount = totalAmount;
+    }
+
+    function validateCollection(
+        address collection,
+        address otcToken,
+        uint256 otcPrice
+    ) external view override returns (bool) {
+        TokenInfo memory info = collections[collection];
+
+        return info.otcToken == otcToken && info.otcPrice == otcPrice;
+    }
+
+    function otc(
+        address collection,
+        uint256 amountToPurchase,
+        uint256 expectedCost
+    ) external payable override onlyExistingCollection(collection) {
+        TokenInfo memory info = collections[collection];
+        uint256 cost = (amountToPurchase * info.otcPrice) / PRICE_MULTIPLIER;
+
+        require(cost == expectedCost, 'OTC: Mismatch to the expected cost');
+
+        if (info.otcToken == address(0)) {
+            require(msg.value == cost, 'OTC: Mismatch to the ETH value');
+        } else {
+            IERC20(info.otcToken).safeTransferFrom(
+                msg.sender,
+                address(this),
+                cost
+            );
+        }
+
+        // Mint HEC
+        hectorTreasury.mintRewards(address(this), amountToPurchase);
```

Comment:
Needs to consider the additional factor of otcPrice because the amountToPurchase alone doesn't factor in the fact that we need an equivalent amount of new Hec equivalent to amountToPurchase * otcPrice.

Reply:
amountToPurchase is the exact amount of HEC we want to buy, so this is okay
otcPrice is the price in otcToken required to purchase exactly 1 HEC